### PR TITLE
Automated build of Windows .exe and .msi artifacts

### PR DESCRIPTION
**Pull request by GregMefford**
* Extract version information from `configure.ac` to `VERSION`
  * Use the WiX Toolset to build an MSI that installs `fwup.exe`
  * Integrate with TravisCI build and release pipeline

### CODE REVIEW

**Comment by fhunleth** [comment]
The static build versions on Travis are failing. There should be a link to the failures on this PR for you, but it's https://travis-ci.org/fhunleth/fwup/builds/153876309 if you don't see it. I'd also make the Windows build separate from the current two Linux builds. When I work with Travis, I usually make 10 or more commits while I debug things and then rebase the branch at the end. There may be a better way, but I haven't found it.

**Comment by GregMefford** [comment]
Travis and I have finally come to terms with this code. :wink:

There could still be issues with the deployment step, since not all the artifacts are built in the same job, but I'm not sure how to test that without going through the release actual process.

File: VERSION [hunk 1]
```diff
@@ -0,0 +1 @@
+0.8.2
```
Comment (fhunleth) [inline VERSION]: Could you change this to 0.9.0-dev? I know, this one was my fault for not changing it earlier...

**Comment by fhunleth** [comment]
Could you enable the regression tests for Windows? I would hope that they running them through wine works as well on Travis as it does on our machines.

**Comment by GregMefford** [comment]
By regression checks, are you just talking about `make check` or is there something else to run? They don't run for static builds on Linux, so I guess it doesn't run for this static Windows build either. Speaking of which, I assume we don't want to build a dynamic-linked version on Windows, right? I think that static is probably a better choice for simplicity on Windows.

**Comment by fhunleth** [comment]
Yes, `make check`. Also, the tests do run (and pass) for the static Linux builds.

Dynamic doesn't make sense for Windows unless we were to target cygwin where the libraries would actually be shared.

**Comment by GregMefford** [comment]
I looked into getting the tests to run for the Windows build and it's going to be surprisingly tricky because the executable is cross-compiled in the Travis VM, but Wine is only available in the Docker container, which doesn't have a proper build environment ready to run the `make check` task. I think it's probably easier to get the required dependencies installed in the Docker container rather than waiting for Wine to build in Travis on each run. I'll keep working toward that.

**Comment by fhunleth** [comment]
"sudo apt-get install wine" doesn't work?

**Comment by GregMefford** [comment]
Hm, good point - that's worth a try. the WiX Toolset requires some .NET stuff that's tricker but I think on my dev box I was able to just use plain `wine`.

**Comment by GregMefford** [comment]
I can't seem to get `wine` to install on Travis because it wants to use an x86 package which [seems to be blocked by Travis](https://github.com/travis-ci/travis-ci/issues/6460). If I can't find a way to get it working in Travis, I'm pretty sure I can get it working in my Docker container, so I'll go that route if I need to.

**Comment by GregMefford** [comment]
Ok, someone on that issue helped me fix the problem in record time, so now it builds and tests using Wine, but it has those same four failing tests that I see on my dev box, related to signing. I believe you said that they don't fail for you under Wine, right? Must be some dependency or environment variable difference.

**Comment by fhunleth** [comment]
Right, those don't fail for me. Could you look into the errors or let me know how they fail when run standalone?

**Comment by GregMefford** [comment]
Using professional-grade printf, I determined that it gets to this line and doesn't come back, when there's a public key specified: https://github.com/GregMefford/fwup/blob/windows/src/fwup_apply.c#L366

I think wine is generating the `abnormal program termination` error below when it fails. The other errors are standard wine warnings that are "normal" debugging output.

```
Application tried to create a window, but no driver could be loaded.
Make sure that your X server is running and that $DISPLAY is set correctly.
err:systray:initialize_systray Could not create tray window
Firmware signing keys created and saved to fwup-key.pub and fwup.priv

Distribute fwup-key.pub with your system so that firmware updates can be
authenticated. Keep fwup.priv in a safe location.
fixme:msvcrt:MSVCRT__wsopen_s : pmode 0x01b6 ignored
fixme:msvcrt:MSVCRT__wsopen_s : pmode 0x01a4 ignored
fixme:msvcrt:MSVCRT__wsopen_s : pmode 0x01a4 ignored

abnormal program termination
```

**Comment by fhunleth** [comment]
Could you use your printf skills to dig a little deeper? https://github.com/GregMefford/fwup/blob/windows/src/cfgfile.c#L424

**Comment by GregMefford** [comment]
Wow, I guess it was late when I stopped digging at that point. For some reason I thought that was a call into `libarchive`, even though I know it's not because I've read through the `cfgfile` code before. :stuck_out_tongue_winking_eye:

Where it actually gets stuck is, somewhat unsurprisingly, the part where it actually tries to call into `libsodium` to use the public key to verify the signature: https://github.com/GregMefford/fwup/blob/windows/src/cfgfile.c#L445

My guess was that maybe libsodium is doing something tricky with hardware-accelerated crypto or something that isn't properly implemented in Wine, which is why I was confused when you said it worked for you.

I'll try to make some time over the weekend to isolate a simpler test case for this function to see if it always fails under Wine, or if it's just something about the way fwup is calling it.

**Comment by fhunleth** [comment]
The issue turned out that weak symbol support in the version of gcc on travis didn't work, and it caused `sodium_memcmp` to crash. I had a newer version of gcc that did work. Luckily, after I tracked it down, I found out that it was fixed upstream in the next version of libsodium.

Also, I fixed a couple minor things and removed .msi file generation. It builds on Travis now and deploys fwup.exe to S3. With any luck it will deploy to GitHub releases when the time comes.

You can get the generated .exe file from http://files.troodon-software.com/list.html.

I merged the changes in this PR to the `windows` branch. Please check the `build_windows.md` file since I wrote it rather quickly.

**Comment by GregMefford** [comment]
Awesome! I'm so excited to get this merged! As we discussed in-person, I'm fine with ditching the `.msi` build for now, since it adds a lot of complexity compared to the value of automating such a simple installation. If someone really wants this functionality in the future, we could easily resurrect it.

**Pull request closed without merging**